lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs

```diff
@@ -168,36 +182,158 @@ data BalanceInsufficientError = BalanceInsufficientError
+        fromMaybe invariantSelectAnyInputs . NE.nonEmpty . UTxOIndex.toList
+
     balanceAvailable :: TokenBundle
     balanceAvailable = view #balance utxoAvailable
 
     balanceRequired :: TokenBundle
-    balanceRequired = F.foldMap (view #tokens) outputsToCover
-
-    mkResult :: SelectionState -> SelectionResult
-    mkResult SelectionState {selected, leftover} =
-        case NE.nonEmpty (UTxOIndex.toList selected) of
-            Nothing ->
-                unableToSelectAnyInputsError
-            Just inputsSelected ->
-                SelectionResult
+    balanceRequired = F.foldMap (view #tokens) outputsConsidered
+
+    -- There's a chicken-and-egg situation when it comes to calculating
+    -- transaction fees. On the one hand, we need to know the shape of the final
+    -- transaction to calculate its cost. But in order to construct the
+    -- transaction, we need to know what its cost is.
+    --
+    -- So, in order to not duplicate the logic from 'makeChange', we first
+    -- calculate a pre-selection considering the case where we have no fees to
+    -- pay, and no minimum value. This is guaranteed to succeed and will yield
+    -- the final shape of the change outputs (modulo amounts).
+    --
+    -- From there, we can successively re-iterate and construct a final
+    -- unbalanced selection that is leaving enough money out for fees. For this
+    -- method to work (efficiently), the function estimate the cost of a
+    -- selection MUST NOT depends on change token quantity. That is, adding more
+    -- change to an output should not make it more expensive.
+    mkPreSelection
+        :: NonEmpty (TxIn, TxOut)
+        -> SelectionResult [AssetId]
+    mkPreSelection inputsSelected =
+        let
+            changeGenerated = maybe
+                (invariantResultWithNoCost inputsSelected)
+                transformChange
+                (makeChange noMinimumCoin noCost
+                    extraCoinSource
+                    (view #tokens . snd <$> inputsSelected)
+                    (view #tokens <$> outputsConsidered)
+                )
+        in
+            SelectionResult
+                { inputsSelected
+                , outputsConsidered
+                , changeGenerated
+                }
+      where
+        noMinimumCoin :: TokenMap -> Coin
+        noMinimumCoin = const (Coin 0)
+
+        noCost :: Coin
+        noCost = Coin 0
+
+        transformChange :: NonEmpty TokenBundle -> NonEmpty [AssetId]
+        transformChange =
+            fmap (fmap fst . snd . TokenBundle.toFlatList)
+
+    -- | This function starts from an initial pre-selection as a way to evaluate
+    -- the cost of a final selection, and then call 'makeChange' repeatedly until
+    -- it succeeds. Between each call, it selects an extra Ada-only input to
+    -- inject additional Ada to construct change address. Eventually it returns
+    -- Just a final selection, or nothing if it runs out of Ada-only inputs.
```

Comment:
Just some minor corrections:

(I'm guessing that "change address" is meant to be "change output"?)

```suggestion
    -- | This function starts from an initial pre-selection as a way to evaluate
    -- the cost of a final selection, and then calls 'makeChange' repeatedly until
    -- it succeeds. Between each call, it selects an extra ada-only input to
    -- inject additional ada to construct change outputs. Eventually it returns
    -- just a final selection, or nothing if it runs out of ada-only inputs.
```